File: LawFirmAPI/LawFirmAPI/Services/CalendarService.cs
// Services/CalendarService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using LawFirmAPI.Data;
using LawFirmAPI.Models.Entities;
using LawFirmAPI.Models.DTOs;

namespace LawFirmAPI.Services
{
    public interface ICalendarService
    {
        Task<List<CalendarEventDto>> GetEvents(long firmId, DateTime? start, DateTime? end);
        Task<CalendarEventDto?> GetEventById(long id, long firmId);
        Task<CalendarEventDto> CreateEvent(long firmId, long userId, CreateCalendarEventDto createDto);
        Task<CalendarEventDto> UpdateEvent(long id, long firmId, CreateCalendarEventDto updateDto);
        Task<bool> DeleteEvent(long id, long firmId);
        Task<List<CalendarEventDto>> GetEventsByDateRange(long firmId, DateTime start, DateTime end);
        Task<List<CalendarEventDto>> GetUpcomingEvents(long firmId, int limit = 10);
        Task<EventAttendeeDto> AddAttendee(long eventId, long firmId, long userId, string attendeeType);
        Task<bool> RemoveAttendee(long eventId, long firmId, long userId);
        Task<EventAttendeeDto> UpdateAttendanceStatus(long eventId, long firmId, long userId, string status);
        Task<EventReminderDto> AddReminder(long eventId, long firmId, DateTime reminderTime, string reminderType);
        Task<bool> RemoveReminder(long reminderId, long firmId);
        Task<CalendarEventDto> SetRecurrence(long eventId, long firmId, EventRecurrenceDto recurrenceDto);
        Task<List<CalendarEventDto>> GetMonthView(long firmId, int year, int month);
        Task<List<CalendarEventDto>> GetWeekView(long firmId, DateTime date);
        Task<List<CalendarEventDto>> GetDayView(long firmId, DateTime date);
        Task<List<CalendarEventDto>> GetAgendaView(long firmId, DateTime start, DateTime end);
    }

    public class CalendarService : ICalendarService
    {
        private readonly ApplicationDbContext _context;

        public CalendarService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<CalendarEventDto>> GetEvents(long firmId, DateTime? start, DateTime? end)
        {
            var query = _context.CalendarEvents
                .Include(e => e.Matter)
                .Include(e => e.Contact)
                .Include(e => e.Attendees)
                .Where(e => e.FirmId == firmId && e.DeletedAt == null);

            if (start.HasValue)
                query = query.Where(e => e.StartDateTime >= start.Value);
            if (end.HasValue)
                query = query.Where(e => e.EndDateTime <= end.Value);

            var events = await query
                .OrderBy(e => e.StartDateTime)
                .Select(e => MapToDto(e))
                .ToListAsync();

            return events;
        }

        public async Task<CalendarEventDto?> GetEventById(long id, long firmId)
        {
            var eventItem = await _context.CalendarEvents
                .Include(e => e.Matter)
                .Include(e => e.Contact)
                .Include(e => e.Attendees)
                .ThenInclude(a => a.User)
                .FirstOrDefaultAsync(e => e.Id == id && e.FirmId == firmId && e.DeletedAt == null);

            return eventItem != null ? MapToDto(eventItem) : null;
        }

        public async Task<CalendarEventDto> CreateEvent(long firmId, long userId, CreateCalendarEventDto createDto)
        {
            var eventItem = new CalendarEvent
            {
                FirmId = firmId,
                Title = createDto.Title,
                Description = createDto.Description,
                Location = createDto.Location,
                EventType = createDto.EventType,
                StartDateTime = createDto.StartDateTime,
                EndDateTime = createDto.EndDateTime,
                IsAllDay = createDto.IsAllDay,
                Color = createDto.Color,
                MatterId = createDto.MatterId,
                ContactId = createDto.ContactId,
                CreatedBy = userId,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _context.CalendarEvents.Add(eventItem);
            await _context.SaveChangesAsync();

            // Add attendees
            if (createDto.AttendeeIds != null)
            {
                foreach (var attendeeId in createDto.AttendeeIds)
                {
                    var attendee = new EventAttendee
                    {
                        EventId = eventItem.Id,
                        UserId = attendeeId,
                        AttendeeType = "REQUIRED",
                        ResponseStatus = "PENDING",
                        CreatedAt = DateTime.UtcNow
                    };
                    _context.EventAttendees.Add(attendee);
                }
                await _context.SaveChangesAsync();
            }

            // Add reminders
            if (createDto.ReminderMinutes.HasValue)
            {
                var reminder = new EventReminder
                {
                    EventId = eventItem.Id,
                    ReminderTime = eventItem.StartDateTime.AddMinutes(-createDto.ReminderMinutes.Value),
                    ReminderType = "BOTH",
                    IsSent = false
                };
                _context.EventReminders.Add(reminder);
                await _context.SaveChangesAsync();
            }

            // Add recurrence
            if (createDto.Recurrence != null)
            {
                await SetRecurrence(eventItem.Id, firmId, createDto.Recurrence);
            }

            return MapToDto(eventItem);
        }

        public async Task<CalendarEventDto> UpdateEvent(long id, long firmId, CreateCalendarEventDto updateDto)
        {
            var eventItem = await _context.CalendarEvents
                .FirstOrDefaultAsync(e => e.Id == id && e.FirmId == firmId && e.DeletedAt == null);

            if (eventItem == null)
                throw new KeyNotFoundException("Event not found");

            eventItem.Title = updateDto.Title;
            eventItem.Description = updateDto.Description;
            eventItem.Location = updateDto.Location;
            eventItem.EventType = updateDto.EventType;
            eventItem.StartDateTime = updateDto.StartDateTime;
            eventItem.EndDateTime = updateDto.EndDateTime;
            eventItem.IsAllDay = updateDto.IsAllDay;
            eventItem.Color = updateDto.Color;
            eventItem.MatterId = updateDto.MatterId;
            eventItem.ContactId = updateDto.ContactId;
            eventItem.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return MapToDto(eventItem);
        }

        public async Task<bool> DeleteEvent(long id, long firmId)
        {
            var eventItem = await _context.CalendarEvents
                .FirstOrDefaultAsync(e => e.Id == id && e.FirmId == firmId && e.DeletedAt == null);

            if (eventItem == null)
                return false;

            eventItem.DeletedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();

            return true;
        }

        public async Task<List<CalendarEventDto>> GetEventsByDateRange(long firmId, DateTime start, DateTime end)
        {
            var events = await _context.CalendarEvents
                .Where(e => e.FirmId == firmId && e.StartDateTime >= start && e.EndDateTime <= end && e.DeletedAt == null)
                .OrderBy(e => e.StartDateTime)
                .Select(e => MapToDto(e))
                .ToListAsync();

            return events;
        }

        public async Task<List<CalendarEventDto>> GetUpcomingEvents(long firmId, int limit = 10)
        {
            var events = await _context.CalendarEvents
                .Where(e => e.FirmId == firmId && e.StartDateTime >= DateTime.UtcNow && e.DeletedAt == null)
                .OrderBy(e => e.StartDateTime)
                .Take(limit)
                .Select(e => MapToDto(e))
                .ToListAsync();

            return events;
        }

        public async Task<EventAttendeeDto> AddAttendee(long eventId, long firmId, long userId, string attendeeType)
        {
            var eventItem = await _context.CalendarEvents
                .FirstOrDefaultAsync(e => e.Id == eventId && e.FirmId == firmId);

            if (eventItem == null)
                throw new KeyNotFoundException("Event not found");

            var existingAttendee = await _context.EventAttendees
                .FirstOrDefaultAsync(a => a.EventId == eventId && a.UserId == userId);

            if (existingAttendee != null)
                throw new InvalidOperationException("User is already an attendee");

            var attendee = new EventAttendee
            {
                EventId = eventId,
                UserId = userId,
                AttendeeType = attendeeType,
                ResponseStatus = "PENDING",
                CreatedAt = DateTime.UtcNow
            };

            _context.EventAttendees.Add(attendee);
            await _context.SaveChangesAsync();

            return new EventAttendeeDto
            {
                UserId = userId,
                AttendeeType = attendeeType,
                ResponseStatus = "PENDING"
            };
        }

        public async Task<bool> RemoveAttendee(long eventId, long firmId, long userId)
        {
            var attendee = await _context.EventAttendees
                .FirstOrDefaultAsync(a => a.EventId == eventId && a.UserId == userId);

            if (attendee == null)
                return false;

            _context.EventAttendees.Remove(attendee);
            await _context.SaveChangesAsync();

            return true;
        }

        public async Task<EventAttendeeDto> UpdateAttendanceStatus(long eventId, long firmId, long userId, string status)
        {
            var attendee = await _context.EventAttendees
                .FirstOrDefaultAsync(a => a.EventId == eventId && a.UserId == userId);

            if (attendee == null)
                throw new KeyNotFoundException("Attendee not found");

            attendee.ResponseStatus = status;
            await _context.SaveChangesAsync();

            return new EventAttendeeDto
            {
                UserId = userId,
                AttendeeType = attendee.AttendeeType,
                ResponseStatus = status
            };
        }

        public async Task<EventReminderDto> AddReminder(long eventId, long firmId, DateTime reminderTime, string reminderType)
        {
            var eventItem = await _context.CalendarEvents
                .FirstOrDefaultAsync(e => e.Id == eventId && e.FirmId == firmId);

            if (eventItem == null)
                throw new KeyNotFoundException("Event not found");

            var reminder = new EventReminder
            {
                EventId = eventId,
                ReminderTime = reminderTime,
                ReminderType = reminderType,
                IsSent = false,
                CreatedAt = DateTime.UtcNow
            };

            _context.EventReminders.Add(reminder);
            await _context.SaveChangesAsync();

            return new EventReminderDto
            {
                Id = reminder.Id,
                ReminderTime = reminderTime,
                ReminderType = reminderType
            };
        }

        public async Task<bool> RemoveReminder(long reminderId, long firmId)
        {
            var reminder = await _context.EventReminders
                .FirstOrDefaultAsync(r => r.Id == reminderId);

            if (reminder == null)
                return false;

            _context.EventReminders.Remove(reminder);
            await _context.SaveChangesAsync();

            return true;
        }

        public async Task<CalendarEventDto> SetRecurrence(long eventId, long firmId, EventRecurrenceDto recurrenceDto)
        {
            var eventItem = await _context.CalendarEvents
                .FirstOrDefaultAsync(e => e.Id == eventId && e.FirmId == firmId);

            if (eventItem == null)
                throw new KeyNotFoundException("Event not found");

            var existingRecurrence = await _context.EventRecurrences
                .FirstOrDefaultAsync(r => r.EventId == eventId);

            if (existingRecurrence != null)
                _context.EventRecurrences.Remove(existingRecurrence);

            var recurrence = new EventRecurrence
            {
                EventId = eventId,
                RecurrencePattern = recurrenceDto.RecurrencePattern,
                IntervalValue = recurrenceDto.IntervalValue,
                DaysOfWeek = recurrenceDto.DaysOfWeek,
                DayOfMonth = recurrenceDto.DayOfMonth,
                MonthOfYear = recurrenceDto.MonthOfYear,
                StartDate = recurrenceDto.StartDate,
                EndDate = recurrenceDto.EndDate,
                Occurrences = recurrenceDto.Occurrences,
                CreatedAt = DateTime.UtcNow
            };

            _context.EventRecurrences.Add(recurrence);
            await _context.SaveChangesAsync();

            return MapToDto(eventItem);
        }

        public async Task<List<CalendarEventDto>> GetMonthView(long firmId, int year, int month)
        {
            var startDate = new DateTime(year, month, 1);
            var endDate = startDate.AddMonths(1).AddDays(-1);

            return await GetEventsByDateRange(firmId, startDate, endDate);
        }

        public async Task<List<CalendarEventDto>> GetWeekView(long firmId, DateTime date)
        {
            var startOfWeek = date.AddDays(-(int)date.DayOfWeek + 1); // Monday
            var endOfWeek = startOfWeek.AddDays(6);

            return await GetEventsByDateRange(firmId, startOfWeek, endOfWeek);
        }

        public async Task<List<CalendarEventDto>> GetDayView(long firmId, DateTime date)
        {
            var startOfDay = date.Date;
            var endOfDay = startOfDay.AddDays(1);

            return await GetEventsByDateRange(firmId, startOfDay, endOfDay);
        }

        public async Task<List<CalendarEventDto>> GetAgendaView(long firmId, DateTime start, DateTime end)
        {
            return await GetEventsByDateRange(firmId, start, end);
        }

        private CalendarEventDto MapToDto(CalendarEvent e)
        {
            return new CalendarEventDto
            {
                Id = e.Id,
                Uuid = e.Uuid,
                Title = e.Title,
                Description = e.Description,
                Location = e.Location,
                EventType = e.EventType,
                StartDateTime = e.StartDateTime,
                EndDateTime = e.EndDateTime,
                IsAllDay = e.IsAllDay,
                Color = e.Color,
                MatterId = e.MatterId,
                MatterTitle = e.Matter?.Title,
                ContactId = e.ContactId,
                ContactName = e.Contact != null ? $"{e.Contact.FirstName} {e.Contact.LastName}" : null,
                CreatedBy = e.CreatedBy,
                Attendees = e.Attendees.Select(a => new EventAttendeeDto
                {
                    UserId = a.UserId,
                    AttendeeType = a.AttendeeType,
                    ResponseStatus = a.ResponseStatus
                }).ToList(),
                CreatedAt = e.CreatedAt,
                UpdatedAt = e.UpdatedAt
            };
        }
    }
}

File: Frontend/law-firm-frontend/src/services/calendar.service.ts
// src/services/calendar.service.ts
import api from './api';
import { CalendarEvent } from '../types';

export const calendarService = {
  // ==================== Basic CRUD Operations ====================
  
  /**
   * Get all events with optional date range
   */
  getEvents: async (start?: string, end?: string): Promise<CalendarEvent[]> => {
    const params: any = {};
    if (start) params.start = start;
    if (end) params.end = end;
    const response = await api.get('/calendar/events', { params });
    return response.data;
  },
  
  /**
   * Get event by ID
   */
  getEventById: async (id: number): Promise<CalendarEvent> => {
    const response = await api.get(`/calendar/events/${id}`);
    return response.data;
  },
  
  /**
   * Create new event
   */
  createEvent: async (data: {
    title: string;
    description?: string;
    location?: string;
    eventType: string;
    startDateTime: string;
    endDateTime: string;
    isAllDay?: boolean;
    color?: string;
    matterId?: number;
    contactId?: number;
    attendeeIds?: number[];
    reminderMinutes?: number;
    recurrence?: {
      recurrencePattern: string;
      intervalValue: number;
      daysOfWeek?: string;
      dayOfMonth?: number;
      monthOfYear?: number;
      startDate: string;
      endDate?: string;
      occurrences?: number;
    };
  }): Promise<CalendarEvent> => {
    const response = await api.post('/calendar/events', data);
    return response.data;
  },
  
  /**
   * Update existing event
   */
  updateEvent: async (id: number, data: {
    title?: string;
    description?: string;
    location?: string;
    eventType?: string;
    startDateTime?: string;
    endDateTime?: string;
    isAllDay?: boolean;
    color?: string;
    matterId?: number;
    contactId?: number;
  }): Promise<CalendarEvent> => {
    const response = await api.put(`/calendar/events/${id}`, data);
    return response.data;
  },
  
  /**
   * Delete event (soft delete)
   */
  deleteEvent: async (id: number): Promise<void> => {
    await api.delete(`/calendar/events/${id}`);
  },
  
  // ==================== Date Range Views ====================
  
  /**
   * Get events by date range
   */
  getEventsByDateRange: async (start: string, end: string): Promise<CalendarEvent[]> => {
    const response = await api.get(`/calendar/events/range?start=${start}&end=${end}`);
    return response.data;
  },
  
  /**
   * Get upcoming events
   */
  getUpcomingEvents: async (limit: number = 10): Promise<CalendarEvent[]> => {
    const response = await api.get(`/calendar/events/upcoming?limit=${limit}`);
    return response.data;
  },
  
  // ==================== View Types ====================
  
  /**
   * Get month view events
   */
  getMonthView: async (year: number, month: number): Promise<CalendarEvent[]> => {
    const response = await api.get(`/calendar/view/month?year=${year}&month=${month}`);
    return response.data;
  },
  
  /**
   * Get week view events
   */
  getWeekView: async (date: string): Promise<CalendarEvent[]> => {
    const response = await api.get(`/calendar/view/week?date=${date}`);
    return response.data;
  },
  
  /**
   * Get day view events
   */
  getDayView: async (date: string): Promise<CalendarEvent[]> => {
    const response = await api.get(`/calendar/view/day?date=${date}`);
    return response.data;
  },
  
  /**
   * Get agenda view events
   */
  getAgendaView: async (start: string, end: string): Promise<CalendarEvent[]> => {
    const response = await api.get(`/calendar/view/agenda?start=${start}&end=${end}`);
    return response.data;
  },
  
  // ==================== Attendee Management ====================
  
  /**
   * Add attendee to event
   */
  addAttendee: async (eventId: number, userId: number, attendeeType: string = 'REQUIRED'): Promise<{
    userId: number;
    attendeeType: string;
    responseStatus: string;
  }> => {
    const response = await api.post(`/calendar/events/${eventId}/attendees?userId=${userId}&attendeeType=${attendeeType}`);
    return response.data;
  },
  
  /**
   * Remove attendee from event
   */
  removeAttendee: async (eventId: number, userId: number): Promise<void> => {
    await api.delete(`/calendar/events/${eventId}/attendees/${userId}`);
  },
  
  /**
   * Update attendee response status
   */
  updateAttendanceStatus: async (eventId: number, userId: number, status: string): Promise<{
    userId: number;
    attendeeType: string;
    responseStatus: string;
  }> => {
    const response = await api.put(`/calendar/events/${eventId}/attendees/${userId}/status?status=${status}`);
    return response.data;
  },
  
  /**
   * Get event attendees
   */
  getEventAttendees: async (eventId: number): Promise<Array<{
    userId: number;
    userName: string;
    userEmail: string;
    attendeeType: string;
    responseStatus: string;
  }>> => {
    const response = await api.get(`/calendar/events/${eventId}/attendees`);
    return response.data;
  },
  
  // ==================== Reminder Management ====================
  
  /**
   * Add reminder to event
   */
  addReminder: async (eventId: number, reminderTime: string, reminderType: string = 'BOTH'): Promise<{
    id: number;
    reminderTime: string;
    reminderType: string;
  }> => {
    const response = await api.post(`/calendar/events/${eventId}/reminders?reminderTime=${reminderTime}&reminderType=${reminderType}`);
    return response.data;
  },
  
  /**
   * Remove reminder from event
   */
  removeReminder: async (reminderId: number): Promise<void> => {
    await api.delete(`/calendar/reminders/${reminderId}`);
  },
  
  /**
   * Get event reminders
   */
  getEventReminders: async (eventId: number): Promise<Array<{
    id: number;
    reminderTime: string;
    reminderType: string;
    isSent: boolean;
  }>> => {
    const response = await api.get(`/calendar/events/${eventId}/reminders`);
    return response.data;
  },
  
  // ==================== Recurrence Management ====================
  
  /**
   * Set recurrence pattern for event
   */
  setRecurrence: async (eventId: number, recurrenceData: {
    recurrencePattern: string;
    intervalValue: number;
    daysOfWeek?: string;
    dayOfMonth?: number;
    monthOfYear?: number;
    startDate: string;
    endDate?: string;
    occurrences?: number;
  }): Promise<CalendarEvent> => {
    const response = await api.put(`/calendar/events/${eventId}/recurrence`, recurrenceData);
    return response.data;
  },
  
  /**
   * Remove recurrence from event
   */
  removeRecurrence: async (eventId: number): Promise<CalendarEvent> => {
    const response = await api.delete(`/calendar/events/${eventId}/recurrence`);
    return response.data;
  },
  
  // ==================== Calendar Integration ====================
  
  /**
   * Connect external calendar (Google, Outlook, etc.)
   */
  connectCalendar: async (data: {
    provider: string;
    accessToken: string;
    refreshToken: string;
    expiresAt: string;
    calendarId?: string;
  }): Promise<{
    provider: string;
    isConnected: boolean;
    calendarId?: string;
  }> => {
    const response = await api.post('/calendar/integrations/connect', data);
    return response.data;
  },
  
  /**
   * Disconnect external calendar
   */
  disconnectCalendar: async (provider: string): Promise<void> => {
    await api.delete(`/calendar/integrations/${provider}`);
  },
  
  /**
   * Get integration status
   */
  getIntegrationStatus: async (): Promise<{
    google: { isConnected: boolean; calendarId?: string; lastSyncAt?: string };
    outlook: { isConnected: boolean; calendarId?: string; lastSyncAt?: string };
  }> => {
    const response = await api.get('/calendar/integrations/status');
    return response.data;
  },
  
  /**
   * Sync with external calendar
   */
  syncCalendar: async (provider: string): Promise<{
    synced: number;
    created: number;
    updated: number;
    deleted: number;
  }> => {
    const response = await api.post(`/calendar/integrations/${provider}/sync`);
    return response.data;
  },
  
  // ==================== Event Types ====================
  
  /**
   * Get all event types
   */
  getEventTypes: async (): Promise<Array<{
    id: string;
    name: string;
    color: string;
    icon: string;
  }>> => {
    const response = await api.get('/calendar/event-types');
    return response.data;
  },
  
  // ==================== Export/Import ====================
  
  /**
   * Export calendar events
   */
  exportEvents: async (format: 'ical' | 'json' | 'csv', start?: string, end?: string): Promise<Blob> => {
    const params: any = { format };
    if (start) params.start = start;
    if (end) params.end = end;
    const response = await api.get('/calendar/export', { params, responseType: 'blob' });
    return response.data;
  },
  
  /**
   * Import events from iCal file
   */
  importEvents: async (file: File): Promise<{
    imported: number;
    failed: number;
    errors: Array<{ row: number; error: string }>;
  }> => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post('/calendar/import', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },
  
  // ==================== Availability ====================
  
  /**
   * Check user availability
   */
  checkAvailability: async (userId: number, start: string, end: string): Promise<{
    available: boolean;
    conflictingEvents: CalendarEvent[];
  }> => {
    const response = await api.get(`/calendar/availability/${userId}?start=${start}&end=${end}`);
    return response.data;
  },
  
  /**
   * Get available time slots
   */
  getAvailableSlots: async (userIds: number[], date: string, durationMinutes: number): Promise<string[]> => {
    const response = await api.post('/calendar/availability/slots', { userIds, date, durationMinutes });
    return response.data;
  },
};